File: apps/windows/TeslaSync.App/feature-views/GForcePanel/GForcePanel.Model.cs
using System.Globalization;
using System.Text.Json;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The mutually-exclusive lifecycle state the <see cref="GForcePanelViewModel"/> can be in — the native union
/// of the branches the web Acceleration-G-Force panel renders
/// (web/src/features/driving/components/driving-dynamics/GForcePanel.tsx). The web component is a thin reader of
/// the <c>useDriveDynamicsLatest</c> snapshot that toggles between the three-up stat grid and a friendly empty
/// state; the native surface binds its own cache-then-network read of <c>/drive-dynamics/latest</c>, so it owns
/// the full loading / loaded / empty / error / stale / offline matrix the P2 state contract requires. Every
/// value maps onto a visible surface (never a blank panel): <see cref="Loaded"/>, <see cref="Stale"/> and
/// <see cref="Offline"/> render the three G-force tiles (with the stale / offline chip for the latter two),
/// <see cref="Empty"/> renders the friendly empty state (web parity: no lateral/longitudinal value yet),
/// <see cref="Loading"/> shows the skeleton chrome and <see cref="Error"/> the retry surface.
/// </summary>
public enum GForcePanelState
{
    /// <summary>Initial fetch with no cached snapshot — render the skeleton chrome.</summary>
    Loading,

    /// <summary>A fresh snapshot carrying at least one acceleration axis.</summary>
    Loaded,

    /// <summary>The snapshot resolved but carries no acceleration value yet.</summary>
    Empty,

    /// <summary>The request failed and no cached snapshot exists — render the retry affordance.</summary>
    Error,

    /// <summary>A cached snapshot older than the freshness window — render the tiles plus a stale chip.</summary>
    Stale,

    /// <summary>The network failed but a cached snapshot remains — render the tiles plus an offline chip.</summary>
    Offline,
}

/// <summary>
/// The parsed <c>/drive-dynamics/latest</c> projection reduced to just the two acceleration axes the web
/// G-force panel reads — the native mirror of the web <c>DriveDynamicsSnapshot</c> subset
/// (<c>lateral_acceleration</c>, <c>longitudinal_acceleration</c>). Field names mirror the Go API's snake_case
/// JSON tags. Parsing is null-tolerant and — exactly like the web's <c>typeof … === 'number'</c> guard — only
/// accepts JSON numbers (a numeric <em>string</em> is treated as absent), so a partial or text-typed payload
/// never produces a misleading reading. WinUI-free so the parse is unit-tested without a UI host.
/// </summary>
public sealed record GForcePanelSnapshot(double? LateralAcceleration, double? LongitudinalAcceleration)
{
    /// <summary>The empty snapshot (no axis reported) — the loading / empty fallback.</summary>
    public static readonly GForcePanelSnapshot Empty = new(null, null);

    /// <summary>True when at least one acceleration axis is present (web <c>hasAny</c>).</summary>
    public bool HasAny => LateralAcceleration is not null || LongitudinalAcceleration is not null;

    /// <summary>Project a single <c>/drive-dynamics/latest</c> JSON object into a tolerant snapshot.</summary>
    public static GForcePanelSnapshot FromJson(JsonElement obj)
    {
        if (obj.ValueKind != JsonValueKind.Object)
        {
            return Empty;
        }

        return new GForcePanelSnapshot(
            GetNumber(obj, "lateral_acceleration"),
            GetNumber(obj, "longitudinal_acceleration"));
    }

    // Web parity: `typeof data?.field === 'number'`. Only a JSON number qualifies; strings / nulls are absent.
    private static double? GetNumber(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var v) || v.ValueKind != JsonValueKind.Number)
        {
            return null;
        }

        return v.TryGetDouble(out var n) && double.IsFinite(n) ? n : null;
    }
}

/// <summary>
/// One projected, render-ready G-force tile — the native mirror of a single web <c>StatCard</c>
/// (icon + label + the formatted value and its <c>g</c> unit). <see cref="Value"/> is the web
/// <c>fmtNumber(_, 2)</c> output or the em-dash fallback; <see cref="Unit"/> is the literal physical symbol
/// <c>g</c> (never localized, matching the web's <c>unit="g"</c>). Pure data so the projection is asserted
/// without a UI host.
/// </summary>
public sealed record GForceMetric(string Label, string Value, string Unit, string AutomationName)
{
    /// <summary>The value and its unit on one line (web baseline-aligned value + unit span, e.g. "0.45 g").</summary>
    public string ValueWithUnit => string.Create(CultureInfo.CurrentCulture, $"{Value} {Unit}");
}

/// <summary>
/// The fully projected, render-ready view of the Acceleration-G-Force surface — the localized title, the three
/// G-force tiles (lateral / longitudinal / combined), the empty-state message and the accessible summary.
/// <see cref="HasData"/> drives the content-vs-empty branch (web parity: the panel shows the grid only when at
/// least one acceleration axis has reported). Pure data so every branch is asserted without a UI host.
/// </summary>
public sealed record GForcePanelDisplay(
    bool HasData,
    string Title,
    IReadOnlyList<GForceMetric> Metrics,
    string EmptyMessage,
    string AriaLabel,
    string AutomationName)
{
    /// <summary>An all-empty display (the friendly empty state) for the loading / empty fallback.</summary>
    public static GForcePanelDisplay Empty(ILocalizer localizer) =>
        GForcePanelProjection.Project(GForcePanelSnapshot.Empty, localizer);
}

/// <summary>
/// Pure projection from a parsed <see cref="GForcePanelSnapshot"/> to a <see cref="GForcePanelDisplay"/> — the
/// native port of the render logic in GForcePanel.tsx. It mirrors the web's per-axis null handling
/// (<c>value != null ? fmtNumber(value, 2) : '—'</c>), the combined-magnitude derivation
/// (<c>sqrt(lateral² + longitudinal²)</c>, only when <em>both</em> axes are present) and the <c>hasAny</c>
/// content-vs-empty gate. Every label resolves through the i18n facade. WinUI-free — unit-tested without a UI
/// host.
/// </summary>
public static class GForcePanelProjection
{
    /// <summary>Fixed display precision (web <c>fmtNumber(value, 2)</c>).</summary>
    public const int Decimals = 2;

    /// <summary>The acceleration unit the tiles carry (web <c>unit="g"</c>, applied verbatim, never localized).</summary>
    public const string Unit = "g";

    /// <summary>The em-dash shown for an absent axis (web <c>'—'</c> fallback).</summary>
    public const string EmDash = "\u2014";

    /// <summary>
    /// The combined acceleration magnitude — the native port of
    /// <c>sqrt(lateral² + longitudinal²)</c>. Returns <c>null</c> unless <em>both</em> axes are present (web
    /// parity: the combined tile shows the em-dash when either axis is missing) or the result is non-finite.
    /// </summary>
    public static double? Magnitude(double? lateral, double? longitudinal)
    {
        if (lateral is not { } lat || longitudinal is not { } lon)
        {
            return null;
        }

        double magnitude = Math.Sqrt((lat * lat) + (lon * lon));
        return double.IsFinite(magnitude) ? magnitude : null;
    }

    /// <summary>Project <paramref name="snapshot"/> using <paramref name="localizer"/> for every label.</summary>
    /// <param name="snapshot">The latest drive-dynamics snapshot (null is treated as no reading).</param>
    /// <param name="localizer">The i18n facade resolving every label.</param>
    public static GForcePanelDisplay Project(GForcePanelSnapshot? snapshot, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);

        string title = localizer.GetString("dynamics.gForce", "Acceleration G-Force");
        string empty = localizer.GetString("dynamics.gForceNoData", "No G-force telemetry received yet");
        string aria = localizer.GetString(
            "dynamics.gForceAria",
            "Acceleration G-force — lateral, longitudinal and combined acceleration in g");

        double? lateral = snapshot?.LateralAcceleration;
        double? longitudinal = snapshot?.LongitudinalAcceleration;
        bool hasAny = lateral is not null || longitudinal is not null;

        if (!hasAny)
        {
            return new GForcePanelDisplay(
                HasData: false,
                Title: title,
                Metrics: Array.Empty<GForceMetric>(),
                EmptyMessage: empty,
                AriaLabel: aria,
                AutomationName: aria);
        }

        var metrics = new[]
        {
            BuildMetric(localizer.GetString("dynamics.lateral", "Lateral"), lateral),
            BuildMetric(localizer.GetString("dynamics.longitudinal", "Longitudinal"), longitudinal),
            BuildMetric(localizer.GetString("dynamics.combined", "Combined"), Magnitude(lateral, longitudinal)),
        };

        return new GForcePanelDisplay(
            HasData: true,
            Title: title,
            Metrics: metrics,
            EmptyMessage: empty,
            AriaLabel: aria,
            AutomationName: aria);
    }

    /// <summary>Format an axis value to the web display contract — <c>fmtNumber(v, 2)</c> or the em-dash.</summary>
    public static string Format(double? value) =>
        value is { } v ? NumberFormatting.Format(v, null, Decimals) : EmDash;

    private static GForceMetric BuildMetric(string label, double? value)
    {
        string text = Format(value);
        string automationName = string.Format(CultureInfo.CurrentCulture, "{0}: {1} {2}", label, text, Unit);
        return new GForceMetric(label, text, Unit, automationName);
    }
}

/// <summary>
/// Maps the engine's raw <c>RepositoryResult&lt;JsonElement&gt;</c> emissions onto parsed
/// <c>RepositoryResult&lt;GForcePanelSnapshot&gt;</c>, preserving every freshness flag (cached / refreshing /
/// stale / offline) so the view-model can render the full state matrix. Pure so the parse-and-preserve
/// contract is unit-tested without a network or cache.
/// </summary>
public static class GForcePanelResultMapper
{
    /// <summary>Parse <paramref name="raw"/>'s payload (when present) while preserving its status.</summary>
    public static RepositoryResult<GForcePanelSnapshot> Map(RepositoryResult<JsonElement> raw)
    {
        ArgumentNullException.ThrowIfNull(raw);

        GForcePanelSnapshot Parse() =>
            raw.HasValue ? GForcePanelSnapshot.FromJson(raw.Value) : GForcePanelSnapshot.Empty;

        return raw.Status switch
        {
            LoadStatus.Loading => RepositoryResult<GForcePanelSnapshot>.Loading(),
            LoadStatus.Cached => RepositoryResult<GForcePanelSnapshot>.Cached(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Refreshing => RepositoryResult<GForcePanelSnapshot>.Refreshing(Parse(), raw.FetchedAt!.Value, raw.IsStale),
            LoadStatus.Loaded => RepositoryResult<GForcePanelSnapshot>.Loaded(Parse(), raw.FetchedAt ?? DateTimeOffset.UtcNow),
            LoadStatus.Empty => RepositoryResult<GForcePanelSnapshot>.Empty(raw.FetchedAt),
            LoadStatus.Offline => RepositoryResult<GForcePanelSnapshot>.OfflineCached(Parse(), raw.FetchedAt!.Value, raw.Error!),
            _ => RepositoryResult<GForcePanelSnapshot>.Failure(
                raw.Error ?? new RepositoryError(RepositoryErrorKind.Unknown, "Unknown error")),
        };
    }
}

/// <summary>
/// Canonical metadata for the Acceleration-G-Force feature surface — the native mirror of the web component at
/// web/src/features/driving/components/driving-dynamics/GForcePanel.tsx. The surface reads the same
/// <c>/drive-dynamics/latest</c> projection the web panel consumes through <c>useDriveDynamicsLatest</c>.
/// </summary>
public static class GForcePanelRegistration
{
    /// <summary>Stable surface id.</summary>
    public const string Id = "g-force-panel";

    /// <summary>Surface category.</summary>
    public const string Category = "driving";

    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "GForcePanel";

    /// <summary>Localized surface name.</summary>
    public static string Name(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer.GetString("dynamics.gForce", "Acceleration G-Force");
    }
}

/// <summary>
/// PII-safe diagnostics for the Acceleration-G-Force surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never a vehicle id or an acceleration value —
/// so a diagnostics line can never leak fleet data. Thread-safe.
/// </summary>
public sealed class GForcePanelDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public GForcePanelDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=GForcePanel</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={GForcePanelRegistration.Slug}");
    }
}

File: apps/apple/TeslaSync/feature-views/GForcePanel.Model.swift
//
//  GForcePanel.Model.swift
//  TeslaSync — P4 feature view · 0169 · GForcePanel (Apple)
//
//  State-holder seam (P1/S8) + telemetry seam (P1/S11 diagnostics) + i18n facade
//  (P1/S10) for the Acceleration G-Force surface. The view binds through
//  `GForcePanelModel`; no networking lives in the view. SwiftUI parity of
//  features/driving/components/driving-dynamics/GForcePanel.tsx — the driving-dynamics
//  panel that shows the live lateral / longitudinal acceleration plus the combined
//  magnitude for the selected vehicle.
//
//  The web source is a thin presentational leaf fed by `useDriveDynamicsLatest`, reading the
//  two acceleration signals (`lateral_acceleration`, `longitudinal_acceleration`) off the
//  `/drive-dynamics/latest` projection and rendering the 3-up stat row (lateral / longitudinal /
//  combined magnitude) when either reading is present (`hasAny`), else its empty state. The native
//  surface owns the full live-query lifecycle through this seam, so the same data the web hook
//  resolves (loading / loaded / empty / failure) plus live-stream freshness (ADR-013 stale /
//  offline) all surface here.
//
//  Vendor-agnostic and SwiftUI-free (Foundation + Observation + OSLog only) so the model + the
//  projection it drives compile and run on a plain host and are pinned by unit tests; the SwiftUI
//  chrome layers on top in GForcePanel.swift / GForcePanel.Views.swift.
//

import Foundation
import Observation
import OSLog

// MARK: - Telemetry seam (P1/S11 diagnostics contract)

/// Emits the `view.opened` product-analytics event for the surface. The default implementation
/// logs via `os.Logger`; the production app injects an adapter that forwards to the shared core
/// `Telemetry.track(.viewOpened(surface:…))`, which is consent-gated and redacted there.
public protocol GForceTelemetry: Sendable {
    func viewOpened(surface: String)
}

/// `os.Logger`-backed default that records the surface open as a `view.opened` event.
public struct OSLogGForceTelemetry: GForceTelemetry {
    private let logger: Logger

    public init() {
        logger = Logger(subsystem: "io.teslasync.app", category: "diagnostics")
    }

    public func viewOpened(surface: String) {
        logger.info("view.opened surface=\(surface, privacy: .public)")
    }
}

// MARK: - State-holder seam (P1/S8 layer)

/// The load lifecycle for the surface's drive-dynamics query, mirroring the shared `LoadableState`
/// cases the web parent projects from its `useDriveDynamicsLatest` hook (web `isLoading` skeleton /
/// resolved snapshot / `data === null` or both-empty accelerations → empty / failure).
public enum GForceLoadStatus: Sendable, Equatable {
    case loading
    case loaded
    case empty
    case failed(String)
}

/// Live-stream freshness (ADR-013). Drives the header freshness chip + the cached-data banner so
/// cached readings are clearly labeled while reconnecting / offline.
public enum GForceConnection: Sendable, Equatable {
    case live
    case stale
    case offline
}

/// The live acceleration reading this surface consumes — the exact subset of the web
/// `DriveDynamicsSnapshot` DTO that `GForcePanel` reads. Lateral acceleration is +ve to the right
/// (cornering) and longitudinal is +ve forward / -ve braking, both expressed in g (the units the
/// `/drive-dynamics/latest` projection emits). Both fields are optional so a partially-populated
/// snapshot projects exactly like the web `typeof === 'number'` guards, and `hasAny` reproduces the
/// web `hasAny` gate that chooses the stat row vs the empty state.
public struct GForceSnapshotInput: Sendable, Equatable {
    public var lateralAcceleration: Double?
    public var longitudinalAcceleration: Double?

    public init(
        lateralAcceleration: Double? = nil,
        longitudinalAcceleration: Double? = nil
    ) {
        self.lateralAcceleration = lateralAcceleration
        self.longitudinalAcceleration = longitudinalAcceleration
    }

    /// Web `hasAny = lateral != null || longitudinal != null` — the gate that decides whether the
    /// stat row renders or the empty state shows.
    public var hasAny: Bool {
        lateralAcceleration != nil || longitudinalAcceleration != nil
    }
}

/// The user's display preferences for this surface. `localeIdentifier` mirrors the global
/// number-format locale `fmtNumber` reads, so grouping + decimal separators match the user's
/// region. The web source formats every g value at a fixed two decimals (`fmtNumber(value, 2)`), so
/// precision is not a user preference here; the view never reads settings directly, the source
/// resolves the locale and pushes it with each snapshot.
public struct GForceUnitPrefs: Sendable, Equatable {
    public var localeIdentifier: String

    public init(localeIdentifier: String = "en_US") {
        self.localeIdentifier = localeIdentifier
    }
}

/// One coalesced snapshot pushed by a `GForceSource`: the live acceleration reading + display prefs
/// plus their load/connection status. The model turns this into the projection + phase.
public struct GForceUpdate: Sendable, Equatable {
    public var status: GForceLoadStatus
    public var connection: GForceConnection
    public var isFetching: Bool
    public var reading: GForceSnapshotInput?
    public var units: GForceUnitPrefs
    public var updatedAt: Date?

    public init(
        status: GForceLoadStatus = .loading,
        connection: GForceConnection = .live,
        isFetching: Bool = false,
        reading: GForceSnapshotInput? = nil,
        units: GForceUnitPrefs = GForceUnitPrefs(),
        updatedAt: Date? = nil
    ) {
        self.status = status
        self.connection = connection
        self.isFetching = isFetching
        self.reading = reading
        self.units = units
        self.updatedAt = updatedAt
    }
}

/// The seam the view binds through. The production app implements this over the shared P1/S8 state
/// holders (`StateHolderModel<LoadableState<DriveDynamicsSnapshot>>` from the KMP drive-dynamics
/// live store composed with the settings store for number formatting); previews and tests use
/// `InMemoryGForceSource`. The view never talks to the network directly.
@MainActor
public protocol GForceSource: AnyObject {
    /// Set by the model; invoked on the main actor for every coalesced snapshot.
    var onUpdate: (@MainActor (GForceUpdate) -> Void)? { get set }
    func start()
    func stop()
    func refresh()
}

/// The surface's observable view-model. Subscribes to a `GForceSource`, recomputes the
/// `GForceProjection` via `GForceProjector`, and exposes a render `Phase` + freshness for SwiftUI to
/// switch over.
@MainActor
@Observable
public final class GForcePanelModel {
    /// The mutually-exclusive render branches: the loading skeleton, the web body's empty branch
    /// (no acceleration signal present → "No G-force telemetry received yet"), a failure (native
    /// retry affordance), and the populated lateral / longitudinal / combined stat row (`hasAny`).
    public enum Phase: Equatable {
        case loading
        case empty
        case error(String)
        case content
    }

    public private(set) var phase: Phase = .loading
    public private(set) var connection: GForceConnection = .live
    public private(set) var isFetching = false
    public private(set) var projection: GForceProjection?
    public private(set) var units = GForceUnitPrefs()
    public private(set) var updatedAt: Date?

    @ObservationIgnored private let source: any GForceSource
    @ObservationIgnored private let telemetry: any GForceTelemetry
    @ObservationIgnored private var started = false
    @ObservationIgnored private var didAutoRefreshForStale = false

    public init(
        source: any GForceSource,
        telemetry: any GForceTelemetry = OSLogGForceTelemetry()
    ) {
        self.source = source
        self.telemetry = telemetry
        source.onUpdate = { [weak self] update in self?.apply(update) }
    }

    /// Begins observing and emits the `view.opened` diagnostics event. Idempotent.
    public func start() {
        guard !started else { return }
        started = true
        telemetry.viewOpened(surface: GForcePanelSurface.slug)
        source.start()
    }

    /// Stops observing the upstream live feed.
    public func stop() {
        started = false
        source.stop()
    }

    /// Forces a refresh (cached readings stay visible). Wired to the retry affordance and to the
    /// stale auto-refresh.
    public func refresh() {
        source.refresh()
    }

    /// Auto-refreshes when the data has gone stale but is not already being fetched — the native
    /// parity of the web stale-query self-refresh (prompt "stale chip + auto-refresh").
    public func autoRefreshIfStale() {
        guard connection == .stale, !isFetching else { return }
        source.refresh()
    }

    private func apply(_ update: GForceUpdate) {
        connection = update.connection
        isFetching = update.isFetching
        units = update.units
        updatedAt = update.updatedAt
        projection = update.reading.map { GForceProjector.project(reading: $0, units: update.units) }
        phase = Self.resolvePhase(status: update.status, hasData: update.reading?.hasAny ?? false)
        handleAutoRefresh(for: update.connection)
    }

    /// Stale → one guarded auto-refresh (prompt "stale chip + auto-refresh"); reset once live so a
    /// later stale episode re-triggers exactly once. Offline keeps the cached reading without
    /// hammering an unreachable backend.
    private func handleAutoRefresh(for connection: GForceConnection) {
        switch connection {
        case .stale:
            guard !didAutoRefreshForStale else { return }
            didAutoRefreshForStale = true
            source.refresh()
        case .live:
            didAutoRefreshForStale = false
        case .offline:
            break
        }
    }

    /// Resolves the render phase. Mirroring the web shell + body: the skeleton shows only on the
    /// initial fetch; the empty state shows when no acceleration signal is present (web `hasAny`
    /// false); whenever a reading is known the stat row renders (cached values stay visible behind a
    /// refresh / transient failure so an offline or stale pod still shows the last-known readings).
    ///
    /// `nonisolated` because it is pure (touches no actor state); this lets the freshness/phase
    /// logic be unit-tested from a non-isolated context under Swift 6 strict concurrency.
    public nonisolated static func resolvePhase(status: GForceLoadStatus, hasData: Bool) -> Phase {
        switch status {
        case .loading:
            hasData ? .content : .loading
        case .empty:
            .empty
        case .loaded:
            hasData ? .content : .empty
        case let .failed(message):
            hasData ? .content : .error(message)
        }
    }
}

/// In-memory source for previews + unit/UI tests. Drive it with `push(_:)`.
@MainActor
public final class InMemoryGForceSource: GForceSource {
    public var onUpdate: (@MainActor (GForceUpdate) -> Void)?
    public private(set) var startCount = 0
    public private(set) var stopCount = 0
    public private(set) var refreshCount = 0

    private let initial: GForceUpdate?

    public init(initial: GForceUpdate? = nil) {
        self.initial = initial
    }

    public func start() {
        startCount += 1
        if let initial { onUpdate?(initial) }
    }

    public func stop() {
        stopCount += 1
    }

    public func refresh() {
        refreshCount += 1
    }

    /// Pushes a snapshot to the bound model (test/preview affordance).
    public func push(_ update: GForceUpdate) {
        onUpdate?(update)
    }
}

// MARK: - Surface identity (kept SwiftUI-free for host-testable model)

/// Diagnostics slug for this surface, kept out of the SwiftUI view so the model/adapter compile and
/// test without SwiftUI. `GForcePanel` re-exposes it as `surfaceSlug` for API parity with the other
/// surfaces.
public enum GForcePanelSurface {
    /// Diagnostics surface slug (P1/S11 `view.opened`).
    public static let slug = "GForcePanel"
}

// MARK: - Localization facade (P1/S10) — web `t(key, default)`

/// Resolves the surface's strings by key with the web English fallback, so the view holds no
/// hardcoded literals. Keys live in the "GForcePanel" table, folded into the app
/// `Localizable.xcstrings` master catalog at integration time; the per-surface table keeps each
/// parallel surface prompt owning its own strings without editing the shared catalog. `string` is
/// Foundation-only so the adapter's accessibility summary can use it; the SwiftUI `text(_:_:)`
/// helper lives in the view file.
public enum GForcePanelStrings {
    public static let table = "GForcePanel"

    public static func string(_ key: String, _ fallback: String) -> String {
        NSLocalizedString(key, tableName: table, bundle: .main, value: fallback, comment: "")
    }
}
